### ISSUE TITLE
UBL character radar code

### ISSUE THREAD

**Issue by Tannoo** [Tannoo/Marlin#7]
Removed

**Comment by Roxy-3D** [comment]
The full MSVC Project is .ZIP'd up and on its way to you...

**Comment by Tannoo** [comment]
Got it.

**Comment by Tannoo** [comment]
Do you have a file that's close to Marlin ready?

If not, I'll plug away at it, but you may end up doing it. lol
I work all weekend, so my time on this will be limited.

**Comment by Tannoo** [comment]
Looks like `Draw_Mesh_Map.cpp` and `LCD_display.cpp`.

**Comment by Roxy-3D** [comment]
Nothing that is ready to plug in and compile....    I have the code fully
debugged...   But it is not adapted to a Marlin environment.

If you keep your progress committed I can pick up from where ever you leave
it and together, it shouldn't take that long to get it going.

**Comment by Roxy-3D** [comment]
ignore LCD_display.cpp    that is only there to give me the ability to use
lcd. type commands....   That entire file goes away.

**Comment by Tannoo** [comment]
`20x4.h` and `20x4-LCD-Mesh-Plot.cpp`?

**Comment by Tannoo** [comment]
It looks as though the main guts are in `Draw_Mesh_Map.cpp`.

I am just trying to figure out what files are the ones to work with.

**Comment by Tannoo** [comment]
Nope, it looks like `20x4-LCD-Mesh-Plot.cpp` and `Draw_Mesh_Map.cpp`.

**Comment by Roxy-3D** [comment]
Yes...  All of the logic is in Draw_Mesh_Map.cpp     There are a couple of
support routines the main routine calls that are also in that file.    But
all the printf()'s  can be remove.   And the several lines of nested loops
to clear the display should be converted to a lcd.clear() or what ever it
is....

And the .h file can be shrunk too...

Incidentally...  If you are on a Windows machine...  You should be able to
run the .exe file that will be in one of the Debug directories.   But if
you just want to look at the output, that is in the tmp.txt file in on of
the Debug directories.

**Comment by Tannoo** [comment]
I tried to run the exe and I got dependency errors. Stuff not installed on my computer.

**Comment by Roxy-3D** [comment]
The main program .cpp file can be discarded too.   That was just there to
set the mesh size to each possible combination, and to call the routine to
plot each mesh point location.

**Comment by Tannoo** [comment]
And that "main" is the one with the main() function in it, I'm sure.

**Comment by Roxy-3D** [comment]
You probably would need to install the MSVC run time library.    But...
You don't need to run it.   You can just look at the tmp.txt file to see
what it does.

**Comment by Tannoo** [comment]
Okay, this is what I have so far:
https://github.com/Tannoo/Marlin/tree/ubl_20x4_radar_integration

It doesn't compile yet. There are some issues with `lcd.customChar()` declarations.
I think I have everything needed.

I created this branch for getting it working because the other branch has an active PR. This one does not.
Once this one compiles and works, I'll apply the changes to the other branch.

**Comment by Roxy-3D** [comment]
> It doesn't compile yet. There are some issues with lcd.customChar() declarations.   I think I have everything needed.

I can get that cleaned up for you pretty easily...   Let me take a look...  And, you may not have noticed...  But I was careful not to have any globally declared custom characters.  And only one on the stack frame that gets used over and over and then discarded when the function exits.  

It now compiles and links cleanly (for a 20x4 LCD Panel).

It is super close.  But somehow the right edge line is getting added to the left part of the mesh point.   I need to write a small routine to dump the number-4 custom character at various points in the routine to see where (and how) it is getting corrupted.   I'll feel more like doing that tomorrow....   But the good news is everything else you have seems to be working.   I can spin the dial and move from pixel to pixel and the nozzle moves where it is supposed to.   So it isn't that far off.

**Comment by Roxy-3D** [comment]
I got it working...  But I don't know how to get your X, Y and Z data along with the Mesh Point index to plot....   I'm going to leave that because I have a hunch it will take all of 5 minutes to get that wired up.

It even lets you click to edit a point.   I think it is very close.    But we probably need to scrub the code and get all the indentation correct.

**Comment by Tannoo** [comment]
Ok.. cool. I'll get the other info on the screen. No big deal. I left them off for simplicity to get the more complicated stuff in there.

**Comment by Tannoo** [comment]
I'll do that in the morning.

**Comment by Tannoo** [comment]
So, does this plotting work down to 16x3?

I ask, so I can get it set appropriately.

**Comment by Tannoo** [comment]
Or just the 16 and 20 x4's?

**Comment by Roxy-3D** [comment]
NO!  Right now UBL has a limit of up to 15 for X & Y.   That simplifies a whole bunch of big data types. 
3x3 to 15x15.   And that tmp.txt file has every mesh pixel plot of what is expected in it.   You are going to see some clever optimizations on  what can be filled in as the dot moves.

**Comment by Roxy-3D** [comment]
> Or just the 16 and 20 x4's?

Its totally untested.   But you should be able to add or subtract characters on any row.   And you should be able to bump it down to 20x2.   (but you would have to put multiple X, Y & Z things on the same line.)

**Comment by Tannoo** [comment]
I just needed to know which displays we want to use it on.
Not the mesh size.

**Comment by Roxy-3D** [comment]
Well...  I have a generic 20 x 4 on that FolgerTech i3-2020.   It is working.   I'm sure there is something that still needs work.  But if you get the X, Y  & Z numbers with the mesh index on the screen...   I'm thinking we merge it.

**Comment by Tannoo** [comment]
Ok. For now, I'll make it only work for the 20x4 lcds. We'll figure out the rest.

One step at a time.

**Comment by Tannoo** [comment]
I should be able to get that and test it in the morning.

**Comment by Roxy-3D** [comment]
The dot moves very nicely across the radar screen.   And the nozzle is pretty responsive.  You can time the speed you rotate the dial to make the nozzle self correct and just head up the center of the bed.

**Comment by Tannoo** [comment]
Yeah.. it's interesting.

Again, I want to thank you for your help on this.

**Comment by Roxy-3D** [comment]
You are very welcome.

**Comment by Tannoo** [comment]
You didn't commit any changes here: https://github.com/Tannoo/Marlin/tree/ubl_20x4_radar_integration

**Comment by Roxy-3D** [comment]
Somehow...  I goofed up and ended up with a branch forked from yours.   All of the changes are at:  https://github.com/Roxy-3D/ubl_20x4_radar_integration      I think we can just take the Ultra Files from that branch and update those same files in your branch.

Lets see if I can fix that...

**Comment by Tannoo** [comment]
I will snag it.

**Comment by Roxy-3D** [comment]
I already moved the stuff to your branch...   I'm wondering if the Radar Map stuff should be moved out of Mesh-Output and to a higher level.  Maybe call it Mesh-Edit ????

**Comment by Tannoo** [comment]
Ok.

**Comment by Roxy-3D** [comment]
Let me know when you have had a chance to check it out....

**Comment by Tannoo** [comment]
Okay... I have and a few corrections to do, if possible.

**Comment by Tannoo** [comment]
The bottom left is drawing correctly, only if the plot is 2 or less on x and y.

**Comment by Tannoo** [comment]
Same with all four corners.

**Comment by Tannoo** [comment]
Lemme get some pics.

**Comment by Tannoo** [comment]
I will also compare again, against your code. Just in case I screwed something up.

**Comment by Roxy-3D** [comment]
What mesh size do you have?  I have 10x10.   I'll load the firmware from your site to make sure we are looking at the same thing.

**Comment by Tannoo** [comment]
Same 10x10.

I have not uploaded my changes yet.... It's working, I am currently formatting it. Close, anyway for now.

**Comment by Roxy-3D** [comment]
OK...  It sounded like there was a problem and you were trying to describe when it didn't work.

**Comment by Tannoo** [comment]
It works, but there is a problem.

**Comment by Tannoo** [comment]
The border characters for the corners.

**Comment by Tannoo** [comment]
Okay, I have committed my changes and it is sync'd.

I'll have to get the pics a little later as I have to get some sleep before work again.

**Comment by Roxy-3D** [comment]
There can only be a 'corner' if the Mesh Point is close to a corner.  Other wise...  It does the best it can.

**Comment by Roxy-3D** [comment]
The reason is...  Two custom char's are used for the top and bottom line.  and two for the left and right line.

And we need 4 char's in certain locations to plot the mesh point.  So that takes up all the custom char's.

**Comment by Tannoo** [comment]
Okay... I thought it was either something I did, or something gone weird.

**Comment by Roxy-3D** [comment]
Nope...   A limitation with how many custom char's we have to work with.    But I do fill in the corner lines if the mesh point is close to a corner!

**Comment by Tannoo** [comment]
But, check it out for you to see if the numbers show up for you. They do for me.

For the 20x4, the radar can be moved to the right 1 or 2 characters.

**Comment by Tannoo** [comment]
Maybe to keep it from looking like a bug... don't fill in those lines... let the pixel do it when it gets to the absolute corners.

**Comment by Roxy-3D** [comment]
Wow!   You did a whole bunch of edits...   I'll load them up.

No matter how it is...  people are going to be offering suggestions to make it look better...   I'm going to just use it the way it is for a while and see how it is...   We can talk about changes later...    And feel free to make any change to fix things that annoy you!!!

**Comment by Tannoo** [comment]
Well, fill in the ones on the left and leave them, if possible.

**Comment by Roxy-3D** [comment]
So you are saying there is a bug???   Or is a change you are suggesting?

**Comment by Tannoo** [comment]
Suggested change.

**Comment by Tannoo** [comment]
Since this box is for the 20x4 only for now... shift the box to the right by 2 characters also.

**Comment by Tannoo** [comment]
There's plenty of room.

**Comment by Roxy-3D** [comment]
OK...   get some sleep....   And when we have more time we can talk about changes...

**Comment by Tannoo** [comment]
You and your coding magic has done it again, though! GREAT WORK!!

**Comment by Roxy-3D** [comment]
That code should have been much easier and simpler to write than it was.   It was one problem after another.   I'm glad I didn't try to write the code directly in Marlin.  The debugging would have been horrible.

**Comment by Roxy-3D** [comment]
Oh!   I see that bug you are talking about...   If the mesh point is on the right side...  the bottom and top line can get extended out when it shouldn't....   I'll have to figure out how that is happening.

**Comment by Tannoo** [comment]
Okay.

**Comment by Tannoo** [comment]
So, aside from formatting, the code sections I added are within the `//==========`'s.

**Comment by Tannoo** [comment]
I changed this:
```cpp
      for (i = 0; i < LCD_WIDTH; i++)
        for (j = 0; j < LCD_HEIGHT; j++) {
          lcd.setCursor(i, j);
          lcd.write(' ');
        }
```
to
```cpp
      lcd_implementation_clear();
```

**Comment by Roxy-3D** [comment]
OK...   I actually had already done that too...   Except I used lcd.clear();

**Comment by Tannoo** [comment]
Maybe lcd.clear() would be faster.

**Comment by Roxy-3D** [comment]
Because that code is in the 20x4 .h file...  You know it is the lcd object.   But for the graphics code, obviously that would not work.    When this code gets moved out of the .h file it really should switch to lcd_implementation_clear();    So probably we should just leave it like you have it.

**Comment by Roxy-3D** [comment]
See if the commit I just did fixes the corner lines when the mesh point is at a corner.   I believe it will...

**Comment by Tannoo** [comment]
I will in the morning. (At work again)

**Comment by Roxy-3D** [comment]
Where do you work?

**Comment by Tannoo** [comment]
At a foundary. I am a maintenance electrician.

**Comment by Tannoo** [comment]
I normally have this weekend off, but someone HAD to go on vacation... lucky me. Lol

**Comment by Roxy-3D** [comment]
It would seem electrical wires shouldn't wear out.   They should just sit there and not break.   You probably aren't very busy because of that.

**Comment by Tannoo** [comment]
Well, there's more to the job than that. Lol
You are correct though, if the wires are secured. Vibration is the killer on all kinds of connection points though.
Other than that, troubleshooting why a machine doesn't run or doesn't do what it's supposed to do.

Sometimes we run accross a bug in the PLC programs also. Rare, but sometimes. Especially after a complete control system upgrade. Hardware and software bugs run rampant then.

**Comment by Tannoo** [comment]
Troubleshooting machines have been a forte' of mine for quite a while now. Not just the 3D printers. I've worked at several different factories.. even a bakery once. I have worked with all kinds of sensors, relays (ssr also), motors, vfds, plcs, and switches. All are basically the same, just different voltage and current capabilities.

C++ / Arduino is fairly new to me. The printer's hardware is also new, but electronics in general are not though. The same does for electronics..all the same, just change of scale.

Did you ask the wrong question? Lol

**Comment by Roxy-3D** [comment]
I was just curious why you always have to work at night...   I knew it was something where the business could not afford any down time.

**Comment by Tannoo** [comment]
Well, we have 4 crews to cover 24/7 operation. We work a 2-2-3 schedule. We also swap days-nights every 6 weeks.

The first half of this year, we have been short on electricians.. so they asked me to stay on nights to help the short screws on nights. Day shift has more support.

They have found a few more guys and I am planned to see daylight once again on the 17th this month. (Shift change).

We will see if I burn or not. Lol

**Comment by Roxy-3D** [comment]
So you must get lots of over time....     Extra pay is always nice.

**Comment by Tannoo** [comment]
Well, some. We have 12hr shifts. So, the 2-2-3 schedule makes it 36 hrs one week and 48 hrs the next. 4hrs OT by schedule. Any personnel coverage is extra OT. This this weekend for me will be 36 hrs OT. (Albiet, split up due to pay cycles)

**Comment by Tannoo** [comment]
I will say though, Uncle Sam love me. Lol

**Comment by Tannoo** [comment]
So... I was looking over the code.

If `upper_left.column = 2`, should move the whole radar over to the right, correct?

If so, we can set an `lcd_width` condition to move the box over.

**Comment by Roxy-3D** [comment]
Go for it!   If we did that, a couple of comments...  Probably the thing to do is very early in the routine when it is trying to figure out everything would be when to determine that.  And the code already does that kind of thing on a 1 pixel basis.   Because the  box line eats up a pixel, if the mesh point size is 5 across or 8 down, it sets an offset flag so it doesn't spill over into the next character cell.

The same thing could be done for this.   And that offset flag could be fed into any lcd.setCursor() call.  (Which there are not that many of them...)    As for lcd_width being a condition, Yes!  But I would also say the number in ULTRA_COLUMNS_FOR_MESH_MAP is a factor.

**Comment by Tannoo** [comment]
I tried `upper_left.column = 2`. That didn't work. I tried up to 11 (thinking in pixels), that didn't work either.

>Probably the thing to do is very early in the routine when it is trying to figure out everything would be when to determine that.

That's what I was thinking.

**Comment by Roxy-3D** [comment]
the .column and .row are in character cells on the LCD display.    The reason is all the routines (like add_edges_to() and such) are trying to determine the position of the custom character they are trying to construct.

**Comment by Tannoo** [comment]
Then a new `box_x_offset` is needed.

**Comment by Tannoo** [comment]
We could leave it like that and I could just move the numbers to the left.

I have just rebased and squashed the branch. You can merge, if you like.

**Comment by Roxy-3D** [comment]
> We could leave it like that and I could just move the numbers to the left.

It might be best to do that...   The reason I say that is because if more stuff is added to the display, it needs to go some where.   And pushing the radar box over the right does look nice, but it eats up screen space.

> I have just rebased and squashed the branch. You can merge, if you like.    

I don't have anything outstanding...

**Comment by Tannoo** [comment]
I just updated my printer with this latest rebased branch...testing it now.

**Comment by Tannoo** [comment]
Okay... don't merge just yet.... that PR branch doesn't have this code yet... give me a few mins.

**Comment by Roxy-3D** [comment]
Oh!    How about we break out the Radar Map to the top level of the UBL Menu and call it 'Mesh Edit' ?    That really is what it does.   And that is the reason people will want to use it.

**Comment by Tannoo** [comment]
You want to call it `Mesh Edit`?

**Comment by Tannoo** [comment]
Just before `UBL Tools`?

**Comment by Roxy-3D** [comment]
If we can think of a better name...  Let's use the better name.   But 'Mesh Edit' is very little text and communicates the idea very well.  I think just before Mesh Storage makes sense.   (It kind of makes sense to have the edit a little sooner and easier to get to than the storage)     But also, I want people to see it.

It can also be in the Mesh Output section.  It doesn't really generate much extra code to have it both places.  (And after all...  That is why you started down this path.  You wanted to be able to spin the dial and see all the mesh values.)   

And maybe we should move the 'Manually Build Mesh' down further in the menu now that people aren't beating on us any more...

**Comment by Tannoo** [comment]
>move the 'Manually Build Mesh' down further in the menu 

Where would you like it?

**Comment by Roxy-3D** [comment]
I see two possibilities...  It could go in Step-by-Step  but the problem with that is Manual probing of the mesh needs to be done only under certain (rare) conditions.  Putting it somewhere like that will just push people to it.

So the other possibility is put it in UBL Tools.   And bury it down in that menu a ways.   Maybe just below  Build Mesh    (Hopefully having Build Mesh followed by Manually Build Mesh will be clear what the distinction is?????)     Actually...   I think this is a 'good' idea.  It makes it clear you can let the machine build most of the mesh.   Or, you can manually build the mesh.    Or you could let the machine build most of the mesh, and then follow up with manual mode for the remaining points....

Your thoughts?

**Comment by Tannoo** [comment]
Okay...sounds good to me.

**Comment by Roxy-3D** [comment]
I just had trouble shutting down and getting out of the Radar Mesh Edit again.   How am I supposed to shut down that function?   If I press and hold the encoder wheel...  It just jumps back in.

**Comment by Tannoo** [comment]
lol... Click to edit... wait for the Z axis to stop moving.. click and hold exit the edit mode.

**Comment by Roxy-3D** [comment]
Yeah...   You are right....   I was just being impatient.   I wonder if anybody else is going to have trouble with that?   Part of my problem is my Z-Axis is so slow on my FolgerTech i3-2020.     There are a couple of ways we could make it so the edit could be aborted quicker if other people have problems 'escaping' from the editor.

**Comment by Tannoo** [comment]
We will see.

Okay.. waiting on travis check. I have to go pass out.

UBL radar map corrections and cleanup #7231
